src/compiler/moduleSpecifiers.ts

```diff
@@ -1026,6 +1028,46 @@ function tryGetModuleNameFromPackageJsonImports(moduleFileName: string, sourceDi
     })?.moduleFileToTry;
 }
 
+function tryGetModuleNameFromPackageJsonDependencies(moduleFileName: Path, { sourceDirectory, getCanonicalFileName }: Info, options: CompilerOptions, host: ModuleSpecifierResolutionHost, importMode: ResolutionMode) {
+    if (!host.readFile) {
+        return undefined;
+    }
+    const cache = host.getModuleResolutionCache?.();
+    if (!cache) {
+        return undefined;
+    }
+
+    const ancestorDirectoryWithPackageJson = getNearestAncestorDirectoryWithPackageJson(host, sourceDirectory);
+    if (!ancestorDirectoryWithPackageJson) {
+        return undefined;
+    }
+    const packageJsonPath = combinePaths(ancestorDirectoryWithPackageJson, "package.json");
+    const cachedPackageJson = host.getPackageJsonInfoCache?.()?.getPackageJsonInfo(packageJsonPath);
+    if (isMissingPackageJsonInfo(cachedPackageJson) || !host.fileExists(packageJsonPath)) {
+        return undefined;
+    }
+    const packageJsonContent = cachedPackageJson?.contents.packageJsonContent || tryParseJson(host.readFile(packageJsonPath)!);
+    const dependencies = packageJsonContent?.dependencies;
+    if (!dependencies) {
+        return undefined;
+    }
+
+    const candidates = getOwnKeys(dependencies);
+    for (const name of candidates) {
+        const resolved = resolveModuleName(name, combinePaths(sourceDirectory, "dummy.ts"), options, host as (typeof host & { readFile: NonNullable<(typeof host)["readFile"]>; }), cache, /*redirectedReference*/ undefined, importMode);

```

Comment:
Should `ModuleSpecifierResolutionHost` have a required `readFile` / be a subtype of `ModuleResolutionHost`? N.B. these functions may be called with hosts from public API usage, though I think `readFile` is safe because it’s required on `LanguageServiceHost`.

Reply:
I mean, this funky-lookin' cast is only here because 1. it's minimal (I could do `as ModuleResolutionHost`, I just feel like this cast more precisely follows the check done in the function body) and 2. there's a `if (!host.readFile) { return }` at the top which doesn't narrow the type as passed in here. Potential future improvement in checking there. But as for

> Should ModuleSpecifierResolutionHost have a required readFile / be a subtype of ModuleResolutionHost?

we could... but I think having it optional reminds us to always prefer to check caches before we go crawling the disk, so we at least have that fallback available wherever possible. It's reasonable if we do want to break the API, I just don't think we need to overmuch, and as it is, it keeps us in good habits.

Reply:
I missed the check for `host.readFile` earlier—no issues with this then 👍